File: MetodosJavascript/Algoritmos/WinogradScaled.js
/*
Algoritmo tomado del artículo https://arxiv.org/abs/1106.1347 y adaptado con la ayuda de inteligencia artificial

Integrantes del grupo:
Esteban Julián Ortega Tapie
Jhon Sebastián Zambrano Ariza
David Serna Restrepo
*/

import { winograd_original } from "./WinogradOriginal.js"

/**
 * Realiza la multiplicación de matrices utilizando el algoritmo de Winograd escalado.
 * 
 * @param {number[][]} matrizA - La primera matriz a multiplicar.
 * @param {number[][]} matrizB - La segunda matriz a multiplicar.
 * @param {number[][]} matrizResultado - Matriz donde se almacenará el resultado de la multiplicación.
 * @param {number} cantidadFilasMatrices - Número de filas en las matrices de entrada.
 * @param {number} cantidadColumnasMatrices - Número de columnas en las matrices de entrada.
 * @param {number} delimitadorMaximaIteracionesFilaColumna - Límite para el número máximo de iteraciones en filas o columnas.
 * @returns {number[][]} - La matriz resultante de la multiplicación.
 */
export function winograd_scaled(matrizA, matrizB, matrizResultado, cantidadFilasMatrices, cantidadColumnasMatrices, delimitadorMaximaIteracionesFilaColumna) {

    // Copia de las matrices originales
    let copiaMatrizA = matrizA.map(fila => fila.map(() => undefined));
    let copiaMatrizB = matrizB.map(fila => fila.map(() => undefined));

    // Cálculo de la norma infinito para las matrices A y B
    let a = norma_infinito(matrizA);
    let b = norma_infinito(matrizB);

    // Cálculo de lambda
    let lambda = Math.floor(0.5 + Math.log(b / a) / Math.log(4));

    // Escalado de las matrices A y B
    copiaMatrizA = multiplicacion_con_escalar(matrizA, matrizResultado, Math.pow(2, lambda));
    copiaMatrizB = multiplicacion_con_escalar(matrizB, matrizResultado, Math.pow(2, -lambda));

    // Aplicación del algoritmo de Winograd original
    matrizResultado = winograd_original(copiaMatrizA, copiaMatrizB, matrizResultado, cantidadFilasMatrices, cantidadColumnasMatrices, delimitadorMaximaIteracionesFilaColumna);
    
    // Devuelve la matriz resultado
    return matrizResultado;
}

/**
 * Calcula la norma infinito de una matriz.
 * 
 * @param {number[][]} matriz - La matriz de entrada.
 * @returns {number} - La norma infinito de la matriz.
 */
function norma_infinito(matriz) {
    let norm = 0.0;
    let aux = matriz.reduce((sum, row) => {
        return sum + row.reduce((rowSum, num) => rowSum + Math.abs(num), 0)
    }, 0);
    norm = Math.max(norm, aux);
    return norm;
}

/**
 * Multiplica una matriz por un escalar.
 * 
 * @param {number[][]} matrizA - La matriz a escalar.
 * @param {number[][]} matrizResultado - Matriz donde se almacenará el resultado de la multiplicación por escalar.
 * @param {number} escalar - El escalar por el cual se multiplica la matriz.
 * @returns {number[][]} - La matriz resultante de la multiplicación por escalar.
 */
function multiplicacion_con_escalar(matrizA, matrizResultado, escalar) {
    matrizResultado = matrizA.map(fila => fila.map(valor => valor * escalar));
    return matrizResultado;
}

File: MetodosPython/Algoritmos/WinogradScaled.py

'''
Algoritmo tomado del artículo https://arxiv.org/abs/1106.1347 y adaptado con la ayuda de inteligencia artificial

Integrantes del grupo:
Esteban Julián Ortega Tapie
Jhon Sebastián Zambrano Ariza
David Serna Restrepo

'''


import numpy
# Importa la función winograd_original del módulo WinogradOriginal

from WinogradOriginal import winograd_original

"""
    Algoritmo de multiplicación de matrices basado en el método de Winograd original.

    Args:
    - matrizA (number[][]): Matriz de entrada A.
    - matrizB (number[][]): Matriz de entrada B.
    - matrizResultado (number[][]): Matriz donde se almacenará el resultado de la multiplicación.
    - cantidadFilasMatrices (int): Cantidad de filas de las matrices de entrada.
    - cantidadColumnasMatrices (int): Cantidad de columnas de las matrices de entrada.
    - delimitadorMaximaIteracionesFilaColumna (int): Delimitador máximo para las iteraciones de fila y columna.

    Returns:
    - (number[][]): Matriz resultado de la multiplicación de A y B.
    """

def winograd_scaled(matrizA, matrizB, matrizResultado, cantidadFilasMatrices, cantidadColumnasMatrices, delimitadorMaximaIteracionesFilaColumna):
    # Realiza una copia de las matrices de entrada

  copiaMatrizA = [fila[:] for fila in matrizA]
  
  copiaMatrizB = [fila[:] for fila in matrizB]

    # Calcula la norma infinito de las matrices de entrada

  a = norma_infinito(matrizA)
  b = norma_infinito(matrizB)
  
  lambda_ = int(0.5 + numpy.log(b/a)/numpy.log(4))

  copiaMatrizA = multiplicacion_con_escalar(matrizA, matrizResultado, 2**lambda_)
  copiaMatrizB = multiplicacion_con_escalar(matrizB, matrizResultado, 2**-lambda_)

  matrizResultado=winograd_original(copiaMatrizA, copiaMatrizB, matrizResultado, cantidadFilasMatrices, cantidadColumnasMatrices, delimitadorMaximaIteracionesFilaColumna)
  return matrizResultado     # Retorna matriz resultado


def norma_infinito(matriz):

  norm = 0.0
    # Calcula la suma de los máximos valores absolutos de cada fila de la matriz y los suma

  aux = sum(max(abs(x) for x in fila) for fila in matriz)
    # Actualiza la norma al máximo entre la norma actual y aux

  norm = max(norm, aux)

  return norm       # Retorna la norma calculada


# Define una función para la multiplicación de matrices con escalares

def multiplicacion_con_escalar(matrizA, matrizResultado, escalar):
  # Multiplica cada elemento de la matrizA por el escalar y lo asigna a matrizResultado

  matrizResultado = [[x * escalar for x in fila] for fila in matrizA]

  return matrizResultado     # Retorna la matriz resultado
